lib/io.h

```diff
@@ -62,6 +62,15 @@ struct metal_io_ops {
 				 memory_order order,
 				 int len);
 	void		(*close)(struct metal_io_region *io);
+	void*		(*offset_to_virt)(struct metal_io_region *io,
+					  unsigned long offset);
+	unsigned long	(*virt_to_offset)(struct metal_io_region *io,
+					  void* virt);
+	metal_phys_addr_t
+			(*offset_to_phys)(struct metal_io_region *io,
+					  unsigned long offset);
+	unsigned long	(*phys_to_offset)(struct metal_io_region *io,
+					  metal_phys_addr_t phys);
```

Comment:
Could you share the use case on why it requires customised offset to virt/phys conversion?

Reply:
Some complex RTOS(e.g. NuttX) has the dedicated subsystem to manage the memory, which:
1.Encapsulate the hardware MMU difference and
2.Expose a set of API to manage the virtual and physical address mapping
So when OpenAMP is ported to this type of RTOS, it's better to reuse the address conversion API provided by RTOS directly, instead duplicating the translation table again.
The callback is a simple way to achieve this flexibility and keep the backward compatibility at the same time.

Reply:
Here is what's not clear to me: the io region assumes that the  virtual is contiguous but the physical is not, the offset is to the start of the virtual address. In this case, the physmap is an array, you can have multiple phys regions for one IO region. Is there any reason you cannot use physmap array to solve your issue?

Thanks,
Wendy

Reply:
Yes, physmap array could solve this problem theoretically, but this require:
1.Expose OS internal structure which isn't good from encapsulation, and OS desginer will challenge me why not use the OS API to do the address conversion directly.
2. Some DSP need the very specific VA to PA mapping which can't represent by physmap array. For example, TL42x(many audio DSP) use word(2 byte) as the base unit, DSP address 0x1000 really point to 0x2000 byte, here we have to multiple 2 for va->pa convesion. With callback function, we can avoid this ugly code in the common code. 
3.Finally, the new callback functions follow other callback design(write/read/block_write/block_read/block_set), I think this is a natural extension.

Reply:
> Yes, physmap array could solve this problem theoretically, but this require:
> 1.Expose OS internal structure which isn't good from encapsulation, and OS desginer will challenge me why not use the OS API to do the address conversion directly.

Agreed
> 2. Some DSP need the very specific VA to PA mapping which can't represent by physmap array. For example, TL42x(many audio DSP) use word(2 byte) as the base unit, DSP address 0x1000 really point to 0x2000 byte, here we have to multiple 2 for va->pa convesion. With callback function, we can avoid this ugly code in the common code.

Try to understand more on this point. your address is 2byptes.
when you get the physical address out of his (DSP address 0x1000 and TL42x 0x2000) region.
do you use this physical address locally or you use it remotely? that is do you need to pass this address to the remote?

We also have remote address which is different to local bus address cases.
Is this one causing the issue?

Is your OpenAMP application runs on TL42x core? if yes, what's the physmap set in your application, dsp (0x1000) or 0x2000?

do you use both "0x1000" and "0x2000" in the same application runs on TL42x core?
Have you also consider two io regions in this case?

> 3.Finally, the new callback functions follow other callback design(write/read/block_write/block_read/block_set), I think this is a natural extension.

Reply:
> > Yes, physmap array could solve this problem theoretically, but this require:
> > 1.Expose OS internal structure which isn't good from encapsulation, and OS desginer will challenge me why not use the OS API to do the address conversion directly.
> 
> Agreed
> 
> > 1. Some DSP need the very specific VA to PA mapping which can't represent by physmap array. For example, TL42x(many audio DSP) use word(2 byte) as the base unit, DSP address 0x1000 really point to 0x2000 byte, here we have to multiple 2 for va->pa convesion. With callback function, we can avoid this ugly code in the common code.
> 
> Try to understand more on this point. your address is 2byptes.
> when you get the physical address out of this (DSP address 0x1000 / 0x2000) region.
> do you use this physical address locally or you use it remotely? that is do you need to pass this address to the remote?
> 
TL42x use 0x1000(VA) locally, but it has to convert to 0x2000(PA) before pass to other side.

> We also have remote address which is different to local bus address cases.
> Is this one causing the issue?
This is another issue we face could resolve by physmap,  but TL42x can't use physmap since the mapping isn't linear. Actually, we extend the virtual address concept:
1.Not only include the address which go through MMU
2.but also include the address go through some bus mapping
Both address need go through io_region va<->pa conversion(either physmap or callback).

> In this case, we handle this mapping out of io_region, we manage da_to_pa/pa_to_da within remoteproc in OpenAMP. If this is the case, we had some discussion before on where to put this mapping, in the end, we put this mapping within remoteproc
> 
As I understand, we need three type of address here: VA, PA and DA:
The first pair VA/PA both come from the same CPU, use in the normal IPC comunication, the conversion is provided by io_region.
The second pair PA/DA is different since DA come from another CPU, but PA come from local, this pair is only used during the elf loading and is provided through remoteproc_mem.
I think both mapping VA<->PA and PA<->DA are needed.

> Is your OpenAMP application runs on TL42x core? if yes, what's the physmap set in your application, dsp (0x1000) or 0x2000?
> 
Yes, OpenAMP run on TL42x core. Conceptly, I set VA to 0x1000 and PA to 0x2000, but I really utilize callback here do the mapping since the table can't express the double operation.

> do you use both "0x1000" and "0x2000" in the same application runs on TL42x core?
> 
No, the most code on TL42x use 0x1000, 0x2000 is only used to set the vring buffer address in the share memory.

> > 3.Finally, the new callback functions follow other callback design(write/read/block_write/block_read/block_set), I think this is a natural extension.

Reply:
> linear. Actually, we extend the virtual address concept:
> 1.Not only include the address which go through MMU
> 2.but also include the address go through some bus mapping
> Both address need go through io_region va<->pa conversion(either physmap or callback).

You have virtual to offset and offset to virtual besides physical address. That is for you io_region, even the logical address va is not linear. Could you clarify that? Do you also need offset <-> va mapping?

Reply:
No, we don't really need offset<->va mapping. Actually, in NuttX porting:
https://github.com/OpenAMP/libmetal/blob/fd213750017660437f440dad3ecb13eded4e3f19/lib/system/nuttx/io.c
you can see the offset<->va is the identity mapping.
this issue is related to the current io_region desgin:
1.va<->pa mapping need convert to offset first
2.read/write/set use offset notation, not virtual address
I have to follow it.
Actually, I think the offset in io_region is redundancy, if we remove offset from io_region and:
1.convert pa->va and va->pa directly
2.read/wreite/set use va(void *) as input
This desgin is much simple than before.

Reply:
> you can see the offset<->va is the identity mapping.
> this issue is related to the current io_region desgin:
> 1.va<->pa mapping need convert to offset first
But shouldn't the offset is the offset  to the base of the va address, should it be just linear?

You just need to do offset <->pa?
> 2.read/write/set use offset notation, not virtual address
I think the offset is the offset to the base of va, and thus, the offset <-> va should not need any changes.

> I have to follow it.
> Actually, I think the offset in io_region is redundancy, if we remove offset from io_region and:
> 1.convert pa->va and va->pa directly
> 2.read/wreite/set use va(void *) as input
> This desgin is much simple than before.

Reply:
You just need to do offset <->pa?
No, the really need is va <->pa. As I emphasize before, the current io_region desgin require:
1.va need first convert to offset
2.offset convert to pa(or call read/write/set)
This is the io_region desgin issue, before io_region remove offset notation from all API/callback, the callback for offset<->va and offset<->pa is required.
Once offset notation is removed, we just need add va<->pa callback.

Reply:
One more comment:
OS(include NuttX) normally provide va<->pa conversion API directly, not va<->offset and offset<->pa conversion.
To work with io_region offset notation, I have to:
1.va<->offset use the identity mapping
2.offset<->pa call OS va<->pa API

Reply:
libmetal has metal_io_virt_to_phys() API which sounds like can keep the common implementation but allow OS specific implementation if required.

Just for your io_region, as offset is the offset to va, and thus, i think in your offset <->phys conversion, you can do NuttX_va_to_pa(io->va + offset), and offset = NuttX_pa_to_va(pa) - va.
Will it work?

Reply:
> libmetal has metal_io_virt_to_phys() API which sounds like can keep the common implementation but allow OS specific implementation if required.

Yes, but it is concrete implementation now, I can't easliy replace with my implemenation.
Unless it change from:
return metal_io_phys(io, metal_io_virt_to_offset(io, virt));
to:
return (*io->ops.virt_to_phys)(io, virt);

> 
> Just for your io_region, as offset is the offset to va, and thus, i think in your offset <->phys conversion, you can do NuttX_va_to_pa(io->va + offset), and offset = NuttX_pa_to_va(pa) - va.
> Will it work?

I can't understand very well the question.
In my case, offset isn't the offset to va, offset actually equal va. So:
1.For va->offset, I just return va as offset
2.For offset->pa, I just call NuttX_va_to_pa(offset)
3.For pa->offset, I just call NuttX_pa_to_va(pa)
4.For offset->va, I just reutrn offset as va

After some thought, I can initialize io->virt=0, io->size=0xffffffff to make metal_io_virt/metal_io_virt_to_offset become the identity mapping, and then remove offset_to_virt/virt_to_offset callback.
Is it your want?